### PR TITLE
Add a base-color uniform

### PR DESCRIPTION
**Pull request by armansito**
This PR introduces a clear-color to apply as the base color for blends in the fine rasterizer stage. This uniform only has an effect in the full fine pipeline.

This PR also includes a minor reorganization of render-time parameters (the new clear color, target image widht/height) in a new `RenderParams` struct and updates the examples accordingly.

### CODE REVIEW

File: examples/with_winit/src/main.rs
Comment (dfrg) [inline examples/with_winit/src/main.rs:68]: The peniko `Color` type has a [parse](https://github.com/linebender/peniko/blob/9526fb1afa85493cf537afac76c1e38dc7c76d23/src/color.rs#L128) method that supports hex values and svg color names. Can we make use of that here instead of a new enum with a limited set of colors?

**Review by dfrg — COMMENTED** [review]
The plumbing and GPU side all looks good! Added a note inline about potentially reusing peniko color parsing for the CLI arg.

File: src/lib.rs
Comment (DJMcNab) [inline src/lib.rs:67]: `Renderer` is never created without 'default options' as far as I can see.

File: src/render.rs
Comment (DJMcNab) [inline src/render.rs:65]: `struct Config` (Rust) and `struct Config` (wgsl) need comments specifying that they must be kept in sync in the correct order.

This is as good a time as any to add them

File: examples/with_winit/src/main.rs
Comment (DJMcNab) [reply examples/with_winit/src/main.rs:62]: A scene which cycles through all clear colours of a specific luminance/saturation would be cool, although that seems equivalent to just doing the same thing with a normal fill.

File: examples/with_bevy/src/main.rs
Comment (DJMcNab) [inline examples/with_bevy/src/main.rs:50]: Is it a clear colour or a base colour? For me, clear colour suggests an extra step in the pipeline, running a clearing step.

It's probably a distinction without a difference

File: examples/with_winit/src/main.rs
Comment (DJMcNab) [inline examples/with_winit/src/main.rs:56]: This is missing some docs

It can probably also reasonably move to `scenes`. It seems equally applicable to `headless`

**Review by DJMcNab — COMMENTED** [review]
Seems like a useful little feature, just some peripheral concerns.

File: examples/with_bevy/src/main.rs
Comment (armansito) [reply examples/with_bevy/src/main.rs:50]: I certainly had the same question as I settled on clear color, as it is slightly misleading if you think of it as the clear blit in a render pass.  Base color sounds good to me.

File: examples/with_winit/src/main.rs
Comment (armansito) [reply examples/with_winit/src/main.rs]: Great! Done.

File: examples/with_winit/src/main.rs
Comment (armansito) [reply examples/with_winit/src/main.rs]: > This should be an Option, because scenes should be able to set the clear colour if not specified by the user.

Good idea! I made this an optional parameter and made it the responsibility of the examples to supply a default color.

> A scene which cycles through all clear colours of a specific luminance/saturation would be cool, although that seems equivalent to just doing the same thing with a normal fill.

I thought of this, though since I defined this as a render parameter (rather than a scene parameter) this would require changing the test_scenes plumbing a little bit so that render options can be modified and animated by the test scenes. Which is OK, and I'll think about it as a follow up.

I considered making this a scene parameter as well but I wasn't sure whether this belongs in the scene or as a renderer option. Maybe there is value in specifying a background color at the scene level; I don't feel strongly either way.

@dfrg @raphlinus Any thoughts?

File: examples/with_winit/src/main.rs
Comment (armansito) [reply examples/with_winit/src/main.rs]: Good idea! Done.

File: src/lib.rs
Comment (armansito) [reply src/lib.rs]: Ah this comment is leftover from the first patch where I had the options at the Renderer level. Removed.

File: src/render.rs
Comment (armansito) [reply src/render.rs]: Good point! Done.

**Comment by armansito** [comment]
Thank you for the feedback everyone. I pushed a new commit and responded inline to all comments. Let me know if I missed anything!

File: src/encoding/packed.rs
Comment (DJMcNab) [inline src/encoding/packed.rs:52]: I think we could do with a "backlink" here, i.e. an equivalent comment referring to this file in `shaders/shared/config.wgsl`

File: src/render.rs
Comment (DJMcNab) [inline src/render.rs:59]: Oh, ouch. The synchronisation is in three places?

**Review by DJMcNab — COMMENTED** [review]
Looks reasonable to me, presuming it runs correctly on web (which I can't test)

File: src/render.rs
Comment (dfrg) [reply src/render.rs:59]: The one in render.rs should go away. I’ll do a cleanup pass soon.

**Review by dfrg — APPROVED** [review]
This looks good to me. Thanks!

File: src/render.rs
Comment (dfrg) [reply src/render.rs:59]: The config struct in packed.rs will also move in the glyph run PR so I’ll fix up the comment there. I think we should land this as is.

File: examples/with_winit/src/main.rs
Comment (DJMcNab) [reply examples/with_winit/src/main.rs]: > Anything that can be parameterized wrt to scene probably should so I prefer the current state.

My proposal would be that the priority is:
1. Provided by user on command line
2. Provided by scene
3. Default for renderer (effectively always black)

The option in 2 is mostly useful for a theoretical future where we use `test_scenes` for e.g. CI tests.
In that case, we'd want to validate that base colours works correctly (e.g. correct colour space/alpha sent through, etc.). So a scene which changes the base colour would therefore make sense.

OTOH, the current state is also fine, as implementing this scene parameter in such a case would be trivial.

Comment (armansito) [reply examples/with_winit/src/main.rs]: I just pushed a new commit that adds a new scene parameter and a test scene that animates a CIE HCL color (`peniko::Color` had a constructor for this, so I just went with it). I implemented the precedence that you specified. Please try it out.

I'll wait for your approval before I merge the change.

File: src/encoding/packed.rs
Comment (armansito) [reply src/encoding/packed.rs:52]: I'll leave things as is per @dfrg since these types are going to be moving around.

File: src/render.rs
Comment (armansito) [reply src/render.rs:59]: sgtm

**Comment by armansito** [comment]
> Looks reasonable to me, presuming it runs correctly on web (which I can't test)

Unfortunately I can't test this on the web either on the current HEAD due to the issues in https://github.com/linebender/vello/issues/276  😕

File: src/lib.rs
Comment (DJMcNab) [inline src/lib.rs:131]:
```suggestion
        self.render_to_texture(device, queue, scene, &target.view, &params)?;
```

File: examples/headless/src/main.rs
Comment (DJMcNab) [inline examples/headless/src/main.rs:98]: This doesn't compile. That is, `headless` also needs to handle the case where this is set in the scene.

Hopefully we can get #250 sorted soon.

File: src/lib.rs
Comment (DJMcNab) [inline src/lib.rs:118]:
```suggestion
```

**Review by DJMcNab — COMMENTED** [review]
I'm happy for this landing once `headless` compiles and the new warnings are dealt with

File: examples/headless/src/main.rs
Comment (armansito) [reply examples/headless/src/main.rs:98]: Oops! Done.

File: src/lib.rs
Comment (armansito) [reply src/lib.rs]: Done.

**Comment by msiglreith** [comment]
Would it make sense to rather read the values from the framebuffer on the beginning of the fine renderer stage? The initial color would be then set via standard clear operations or could be also some already rendered content.

File: examples/scenes/src/lib.rs
Comment (DJMcNab) [inline examples/scenes/src/lib.rs:53]: The one issue with this is that in powershell, `#` starts a comment. For example:
```powershell
cargo run ... --base-color #bed
```
needs to be written as (with the `` ` `` as the escape character)
```powershell
cargo run ... --base-color `#bed
```

That's not really the fault of this PR though. I don't think we need to add a comment here

**Review by DJMcNab — APPROVED** [review]
Nothing blocking, but some nice things to clean up. Note that we follow the policy 'documented' in https://github.com/linebender/druid/pull/2357#issuecomment-1426814447 here, so you should be the one to merge this PR.

File: examples/with_winit/src/main.rs
Comment (armansito) [reply examples/with_winit/src/main.rs:265]: I wasn't too happy with this. The only reason I put it here was to propagate the error but it's probably not egregious to just unwrap the result inside `run`.

**Comment by armansito** [comment]
> Would it make sense to rather read the values from the framebuffer on the beginning of the fine renderer stage? The initial color would be then set via standard clear operations or could be also some already rendered content.

This is a neat idea and something I have considered. I personally find that the fine rasterization stage having its own base blend color is the right abstraction and generally works for the headless case as well (where there may be no clear color or the client maybe simply starts with an empty texture).

What would be nice as a future improvement is the ability to set the base color for fine with a 0 alpha value and have the blit pass perform an alpha blend over the existing contents of the target texture.

File: examples/scenes/src/lib.rs
Comment (armansito) [reply examples/scenes/src/lib.rs:53]: I would agree that this is a shell-specific concern.

File: examples/with_winit/src/main.rs
Comment (armansito) [reply examples/with_winit/src/main.rs]: > I just want to check ,clap doesn't handle Option<Color> automatically, does it? I think they use FromStr, so it might just work.

Actually this does work pretty well. I used `value_parser` instead of FromStr though I imagine that would work too. The output looks even nicer now. PTAL.

**Pull request merged**
